### ISSUE TITLE
Neon compiler isn't ready for NEO3 contract deployment

### ISSUE THREAD

**Issue by superboyiii** [neo-project/neo-devpack-dotnet#91]
Find many issues need to be fixed in neon complier since it doesn't sync to the latest  neo3 vm.
It's necessary for deploying contract on neo3.
![image](https://user-images.githubusercontent.com/38493437/63493792-f26cfc00-c4ee-11e9-95c4-9371189994d4.png)

**Comment by erikzhang** [comment]
I can compile it well.

**Comment by superboyiii** [comment]
That's great!

**Comment by lock9** [comment]
@superboyiii  can you double check? I believe it has been fixed (I used it in Shanghai). I was able to compile and deploy a simple smart contract without issues.

**Comment by superboyiii** [comment]
> @superboyiii can you double check? I believe it has been fixed (I used it in Shanghai). I was able to compile and deploy a simple smart contract without issues.

Hi, Rick. I made a simple contract and complie it successfully within neon.dll but it isn't able to be deployed on neo-cli. Here's my contract content:
```using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using Neo.SmartContract.Framework.Services.System;
using System;
using System.ComponentModel;
using System.Numerics;

namespace Neo.SmartContract
{
    public class ICO_Template : Framework.SmartContract
    {
        //Token Settings
        public static string Name() => "ZZH";
        public static string Symbol() => "ZZH";
        

        public static Object Main(string operation, params object[] args)
        {
             if (Runtime.Trigger == TriggerType.Application)
            {
               
                if (operation == "name") return Name();
                if (operation == "symbol") return Symbol();
                
            }
            return false;
        }
    }
}```
Could you try it on your local environment?
By the way, I'm using SmartContract.Framework v2.9.3.1 and using the latest neo-devpack-dotnet

**Comment by shargon** [comment]
Deploy require the manifest, we need to upgrade the compiler for generate both, abi and manifest
So in neo-cli, the second param must be a valid json (manifest)

**Comment by superboyiii** [comment]
> Deploy require the manifest, we need to upgrade the compiler for generate both, abi and manifest
> So in neo-cli, the second param must be a valid json (manifest)

How could I turn sc dll to manifest json? I could only get nef file and abi.json from compiler.

**Comment by shargon** [comment]
@superboyiii please, take a look to these pull request

https://github.com/neo-project/neo-devpack-dotnet/pull/109
https://github.com/neo-project/neo-cli/pull/463

**Comment by superboyiii** [comment]
@shargon  Now I tried to deploy a nep-5 demo on NEO3. I find some issues.

1. I wrote a nep5 sc based on the latest sc framework 2.9.3.0
```
using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using Neo.SmartContract.Framework.Services.System;
using System;
using System.ComponentModel;
using System.Numerics;

namespace NEP5
{
    public class NEP5 : SmartContract
    {
        [DisplayName("transfer")]
        public static event Action<byte[], byte[], BigInteger> Transferred;

        private static readonly byte[] Owner = "Ad9R9jPpT8C8gaiDs7e2EZs3mfkrrotCgy".ToScriptHash(); //Owner Address
        private static readonly BigInteger TotalSupplyValue = 10000000000000000;

        public static object Main(string method, object[] args)
        {
            if (Runtime.Trigger == TriggerType.Verification)
            {
                return Runtime.CheckWitness(Owner);
            }
            else if (Runtime.Trigger == TriggerType.Application)
            {
                if (method == "balanceOf") return BalanceOf((byte[])args[0]);

                if (method == "decimals") return Decimals();

                if (method == "deploy") return Deploy();

                if (method == "name") return Name();

                if (method == "symbol") return Symbol();

                if (method == "supportedStandards") return SupportedStandards();

                if (method == "totalSupply") return TotalSupply();

                if (method == "transfer") return Transfer((byte[])args[0], (byte[])args[1], (BigInteger)args[2]);
            }
            return false;
        }

        [DisplayName("balanceOf")]
        public static BigInteger BalanceOf(byte[] account)
        {
            if (account.Length != 20)
                throw new InvalidOperationException("The parameter account SHOULD be 20-byte addresses.");
            StorageMap asset = Storage.CurrentContext.CreateMap(nameof(asset));
            return asset.Get(account).AsBigInteger();
        }
        [DisplayName("decimals")]
        public static byte Decimals() => 8;

        private static bool IsPayable(byte[] to)
        {
            var c = Blockchain.GetContract(to);
            return c == null || c.IsPayable;
        }

        [DisplayName("deploy")]
        public static bool Deploy()
        {
            if (TotalSupply() != 0) return false;
            StorageMap contract = Storage.CurrentContext.CreateMap(nameof(contract));
            contract.Put("totalSupply", TotalSupplyValue);
            StorageMap asset = Storage.CurrentContext.CreateMap(nameof(asset));
            asset.Put(Owner, TotalSupplyValue);
            Transferred(null, Owner, TotalSupplyValue);
            return true;
        }

        [DisplayName("name")]
        public static string Name() => "SUPERBOYIII"; //name of the token

        [DisplayName("symbol")]
        public static string Symbol() => "Owen"; //symbol of the token

        [DisplayName("supportedStandards")]
        public static string[] SupportedStandards() => new string[] { "NEP-5", "NEP-7", "NEP-10" };

        [DisplayName("totalSupply")]
        public static BigInteger TotalSupply()
        {
            StorageMap contract = Storage.CurrentContext.CreateMap(nameof(contract));
            return contract.Get("totalSupply").AsBigInteger();
        }


        // function that is always called when someone wants to transfer tokens.
        // 流转token调用
        [DisplayName("transfer")]
        public static bool Transfer(byte[] from, byte[] to, BigInteger value)
        {
            if (value <= 0) return false;
            if (!Runtime.CheckWitness(from)) return false;
            if (to.Length != 20) return false;

            BigInteger from_value = Storage.Get(Storage.CurrentContext, from).AsBigInteger();
            if (from_value < value) return false;
            if (from == to) return true;
            if (from_value == value)
                Storage.Delete(Storage.CurrentContext, from);
            else
                Storage.Put(Storage.CurrentContext, from, from_value - value);
            BigInteger to_value = Storage.Get(Storage.CurrentContext, to).AsBigInteger();
            Storage.Put(Storage.CurrentContext, to, to_value + value);
            Transferred(from, to, value);
            return true;
        }
    }
}
```

2. Compile it and get nef file and manifest.json file 
![image](https://user-images.githubusercontent.com/38493437/65233534-8ee1e880-db05-11e9-9063-075add8e7100.png)

3. deploy it   successfully and invoke "deploy" method successfully, too. But I'm not able to transfer it.
![image](https://user-images.githubusercontent.com/38493437/65233583-a8833000-db05-11e9-899d-209f2f32b17e.png)
AssetDescriptor return Fault when execution.
![image](https://user-images.githubusercontent.com/38493437/65233632-b9cc3c80-db05-11e9-9605-a0d2079695ca.png)

4. invoke "getbalance" rpc, the asset is truly there.
![image](https://user-images.githubusercontent.com/38493437/65233668-c9e41c00-db05-11e9-974f-2624f02756d1.png)

5.Invoke totalSupply it returns correct value.
![image](https://user-images.githubusercontent.com/38493437/65233707-d8cace80-db05-11e9-89e2-698d0b008249.png)

6.I could also see the eventlog of this deploy
![image](https://user-images.githubusercontent.com/38493437/65233804-09ab0380-db06-11e9-9cc1-a4bb94306311.png)

7. However, invoke ```getnep5transfers```, there's no record expect native nep5 asset.
![image](https://user-images.githubusercontent.com/38493437/65233736-e84a1780-db05-11e9-9970-9e79cdf51797.png)

8. Also no record of ```receive``` from ```getnep5transfers``` expect native nep5 asset.
![image](https://user-images.githubusercontent.com/38493437/65233768-f861f700-db05-11e9-9cce-66009f6651fa.png)
---------------------------------------------------------------------------------------------------------------
In summary, there're could be two issue.
1. Isn't able to invoke ```transfer``` in this nep-5 sc, which returns fault when execution in AssetDescriptor.
2. Nep5Tracker isn't able to catch the balance and transfers related to this asset.

**Comment by shargon** [comment]
A lot of syscalls are pending to be  updated.  

Please check https://github.com/neo-project/neo-devpack-dotnet/pull/106 if you are using one of them

**Comment by superboyiii** [comment]
> A lot of syscalls are pending to be updated.
> 
> Please check #106 if you are using one of them

Wow, got it.
